src/webview/markdownEngine.ts

```diff
@@ -0,0 +1,99 @@
+        } catch (error) {
+            leetCodeChannel.appendLine("[Error] Fail to load built-in markdown style file.");
+            return [];
+        }
+    }
+
+    public get options(): MarkdownIt.Options {
+        return (this.engine as any).options;
+    }
+
+    public getStylesHTML(): string {
+        return this.styles.map((style: vscode.Uri) => `<link rel="stylesheet" type="text/css" href="${style.toString()}">`).join("\n");
+    }
+
+    public render(md: string, env?: any): string {
+        return this.engine.render(md, env);
+    }
+
+    private initEngine(): MarkdownIt {
+        const md: MarkdownIt = new MarkdownIt({
+            linkify: true,
+            typographer: true,
+            highlight: (code: string, lang?: string): string => {
+                if (lang && ["tsx", "typescriptreact"].indexOf(lang.toLocaleLowerCase()) >= 0) {
+                    lang = "jsx";
+                }
+                if (lang && lang.toLocaleLowerCase() === "python3") {
+                    lang = "python";
+                }
+                if (lang && lang.toLocaleLowerCase() === "c#") {
+                    lang = "cs";
+                }
+                if (lang && lang.toLocaleLowerCase() === "json5") {
+                    lang = "json";
+                }
+                if (lang && hljs.getLanguage(lang)) {
+                    try {
+                        return hljs.highlight(lang, code, true).value;
+                    } catch (error) { /* do not highlight */ }
+                }
+                return ""; // use external default escaping
+            },
+        });
+
+        this.addCodeBlockHighlight(md);
+        this.addLinkValidator(md);
+        return md;
+    }
+
+    private addCodeBlockHighlight(md: MarkdownIt): void {
+        const origin: MarkdownIt.TokenRender = md.renderer.rules["code_block"];
+        // tslint:disable-next-line:typedef
+        md.renderer.rules["code_block"] = (tokens, idx, options, env, self) => {
+            // if any token uses lang-specified code fence, then do not highlight code block
+            if (tokens.some((token: any) => token.type === "fence")) {
+                return origin(tokens, idx, options, env, self);
+            }
+            // otherwise, highlight with undefined lang, which could be handled in outside logic.
+            const highlighted: string = options.highlight(tokens[idx].content, undefined);
+            return [
+                `<pre><code ${self.renderAttrs(tokens[idx])} >`,
+                highlighted || md.utils.escapeHtml(tokens[idx].content),
+                "</code></pre>",
+            ].join("\n");
+        };
+    }
+
+    private addLinkValidator(md: MarkdownIt): void {
+        const validateLink: (link: string) => boolean = md.validateLink;
+        md.validateLink = (link: string): boolean => {
+            // support file:// protocal link
+            return validateLink(link) || link.startsWith("file:");
```

Comment:
Could you explain more why we need the validator here?

Reply:
This is the routine adopted by built-in markdown exntesion's [MarkdownEngine](https://github.com/Microsoft/vscode/blob/master/extensions/markdown-language-features/src/markdownEngine.ts#L252), I think it may enables us to validate local file link, which may be useful in local debugger.

Reply:
Ok... Actually still not quite understand... I think it's a worth investigating topic. It's fine to leave it here in this PR...

Reply:
Document from markdown-it:

> MarkdownIt#validateLink(url)Boolean

Link validation function. CommonMark allows too much in links. By default we disable `javascript:`, `vbscript:`, `file:` schemas, and almost all `data:...` schemas except some embedded image types.

You can change this behaviour:
```javascript
var md = require('markdown-it')();
// enable everything
md.validateLink = function () { return true; }
```

Since we will deal with local files in WebView, chances are that it will be helpful to enable `file:` link.

Reply:
I looked into the source code, and found that it was made to prevent XSS attack. Here are some examples:
* When parsing image and link:
```javascript
    if (res.ok) {
      href = state.md.normalizeLink(res.str);
      if (state.md.validateLink(href)) {
        pos = res.pos;
      } else {
        href = '';
      }
    }
```
If `validateLink` does not pass, then `href` is screened out.

* The same goes on for reference-link or auto link:
```js
  href = state.md.normalizeLink(res.str);
  if (!state.md.validateLink(href)) { return false; }
```

So, it is indeed necessary to loose the restriction to allow `file://` protocol.